server/src/main/java/org/elasticsearch/action/admin/cluster/stats/LongMetric.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.cluster.stats;
+
+import org.HdrHistogram.ConcurrentHistogram;
+import org.HdrHistogram.Histogram;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import java.util.zip.DataFormatException;
+
+/**
+ * Metric class that accepts longs and provides count, average, max and maybe percentiles.
+ * Abstracts out the details of how exactly the values are stored and calculated.
+ * {@link LongMetricValue} is a snapshot of the current state of the metric.
+ */
+public class LongMetric {
+    private final Histogram values;
+    private static final int SIGNIFICANT_DIGITS = 2;
+
+    LongMetric() {
+        values = new ConcurrentHistogram(SIGNIFICANT_DIGITS);
+    }
+
+    void record(long v) {
+        values.recordValue(v);
+    }
+
+    LongMetricValue getValue() {
+        return new LongMetricValue(values);
+    }
+
+    /**
+     * Snapshot of {@link LongMetric} value that provides the current state of the metric.
+     * Can be added with another {@link LongMetricValue} object.
+     */
+    public static final class LongMetricValue implements Writeable {
+        // We have to carry the full histogram around since we might need to calculate aggregate percentiles
+        // after collecting individual stats from the nodes, and we can't do that without having the full histogram.
+        // This costs about 2K per metric, which was deemed acceptable.
+        private final Histogram values;
+
+        public LongMetricValue(Histogram values) {
+            // Copy here since we don't want the snapshot value to change if somebody updates the original one
+            this.values = values.copy();
+        }
+
+        public LongMetricValue(LongMetricValue v) {
+            this.values = v.values.copy();
+        }
+
+        LongMetricValue() {
+            this.values = new Histogram(SIGNIFICANT_DIGITS);
+        }
+
+        public void add(LongMetricValue v) {
+            this.values.add(v.values);
+        }
+
+        public static LongMetricValue fromStream(StreamInput in) throws IOException {
+            byte[] b = in.readByteArray();
+            ByteBuffer bb = ByteBuffer.wrap(b);
```

Comment:
What do you think about renaming 'b', 'bb' to something clearer? (byteArray, byteBuffer?)